doc/design/dist/README.md

```diff
@@ -0,0 +1,173 @@
+
+### Parameter Server Process
+
+Parameter server processes hold the parameters collabratively. The parameters are sharded on different parameter servers.
+
+The parameter server will:
+
+- Receive gradient from the trainers, update its parameters, and give the trainers the latest parameters.
+- Periodically save its parameters to distributed file system by overriding the previous save.
+
+### Optimization Algorithms
+
+The communication pattern between the trainers and the parameter servers depends on the category of optimization algorithm:
+
+- Synchronous Stochastic Gradient Decent (sync-SGD)
+
+	Parameter server will wait for all trainer finish n-th mini-batch calculation and send their gradients before broadcasting new parameters to every trainer. Every trainer will wait for the new parameters before starting n+1-th mini-batch.
+  
+- Asynchronous Stochastic Gradient Decent (async-SGD)
+
+	There will no synchronization between different trainers, and parameter server updates its parameter as soon as it receives new gradient:
+
+	- Each trainer uploads its accumulated gradient every **n** mini-batches.
+	- Every **m** mini-batches, the trainer downloads new parameters from parameter server.
+	- **n** and **m** do not have to be equal.
+
+## Fault Tolerant
+
+The training job will pause if the master process is dead, or any of the parameter server process is dead. They will be started by [Kubernetes](https://kubernetes.io/) and recover in few minutes. Please refer to [fault recovery](#fault-recovery).
+
+The training job will continue to make progress if there is at least one training process running. The strategy depends on the type of optimization algorithm:
+
+- sync-SGD
+
+	TODO
+
+- async-SGD
+
+	Since async-SGD does not require synchronization between mini-batches, the system will by definition make process if at least one trainer is running.
+
+## Fault Recovery
+
+PaddlePaddle uses [etcd](https://github.com/coreos/etcd) to keep track of the states of processes. Because etcd is a distributed reliable key-value store, the restarted process can recover its states from etcd. The model parameter are periodically saved into distributed file system, so a restarted parameter server can recover its parameters from the saved file.
+
+Now we will introduce how each process recovers from failure, the graph below provides an illustration:
+
+<img src="src/paddle-etcd.png"/>
+
+### Master Process
+
+When the master is started by the Kubernetes, it executes the following steps at startup:
+
+1. Grabs a unique *master* lock in etcd, which prevents concurrent master instantiations.
+1. Recovers the task queues from etcd if they already exists, otherwise the master will create them.
+1. Watches the trainer prefix keys `/trainer/` on etcd to find the live trainers.
+1. Starts dispatching the tasks to the trainers.
+
+The master process will kill itself if its etcd lease expires.
```

Comment:
@xiang90  thanks! Very good point.

I was planning to create a single [Lease](https://godoc.org/github.com/coreos/etcd/clientv3#Lease) and use it with KV.Put and Mutex.Lock (Mutex is created with a Session with the lease.).

So I suppose in this way I don't have to check the ownership of the lock in the same transaction when it modifies the etcd state (hoping this way could make the code simpler).

The assumption is KV.Put will fail if lease expires, and the master have the lock if lease is not expired -> KV.Put will success only if the master have the lock.
Is my assumption correct?
Thanks!